Source/Particles/Resampling/VelocityCoincidenceThinning.cpp

```diff
@@ -0,0 +1,272 @@
+#if defined(WARPX_DIM_XZ) || defined(WARPX_DIM_3D)
+    auto * const AMREX_RESTRICT x = soa.GetRealData(PIdx::x).data();
+#elif defined(WARPX_DIM_RZ)
+    auto * const AMREX_RESTRICT x = soa.GetRealData(PIdx::x).data(); // rename to PIdx::r after PR #4667
+#endif
+#if defined(WARPX_DIM_3D)
+    auto * const AMREX_RESTRICT y = soa.GetRealData(PIdx::y).data();
+#endif
+    auto * const AMREX_RESTRICT z = soa.GetRealData(PIdx::z).data();
+    auto * const AMREX_RESTRICT ux = soa.GetRealData(PIdx::ux).data();
+    auto * const AMREX_RESTRICT uy = soa.GetRealData(PIdx::uy).data();
+    auto * const AMREX_RESTRICT uz = soa.GetRealData(PIdx::uz).data();
+    auto * const AMREX_RESTRICT w = soa.GetRealData(PIdx::w).data();
+    auto * const AMREX_RESTRICT idcpu = soa.GetIdCPUData().data();
+
+    // Using this function means that we must loop over the cells in the ParallelFor.
+    auto bins = ParticleUtils::findParticlesInEachCell(lev, pti, ptile);
+
+    const auto n_cells = static_cast<int>(bins.numBins());
+    auto *const indices = bins.permutationPtr();
+    auto *const cell_offsets = bins.offsetsPtr();
+
+    const auto min_ppc = m_min_ppc;
+
+    const auto mass = pc->getMass();
+
+    // create a GPU vector to hold the momentum cluster index for each particle
+    amrex::Gpu::DeviceVector<int> momentum_bin_number(bins.numItems());
+    auto* momentum_bin_number_data = momentum_bin_number.data();
+
+    // create a GPU vector to hold the index sorting for the momentum bins
+    amrex::Gpu::DeviceVector<int> sorted_indices(bins.numItems());
+    auto* sorted_indices_data = sorted_indices.data();
+
+    const auto Ntheta = m_ntheta;
+    const auto Nphi = m_nphi;
+
+    const auto dr = m_delta_ur;
+    const auto dtheta = 2.0_prt * MathConst::pi / Ntheta;
+    const auto dphi = MathConst::pi / Nphi;
+    constexpr auto c2 = PhysConst::c * PhysConst::c;
+
+    auto heapSort = HeapSort();
+
+    // Loop over cells
+    amrex::ParallelForRNG( n_cells,
+        [=] AMREX_GPU_DEVICE (int i_cell, amrex::RandomEngine const& engine) noexcept
+        {
+            // The particles that are in the cell `i_cell` are
+            // given by the `indices[cell_start:cell_stop]`
+            const auto cell_start = static_cast<int>(cell_offsets[i_cell]);
+            const auto cell_stop  = static_cast<int>(cell_offsets[i_cell+1]);
+            const auto cell_numparts = cell_stop - cell_start;
+
+            // do nothing for cells with less particles than min_ppc
+            // (this intentionally includes skipping empty cells, too)
+            if (cell_numparts < min_ppc) {
+                return;
+            }
+
+            // Loop over particles and label them with the appropriate momentum bin
+            // number. Also assign initial ordering to the sorted_indices array.
+            for (int i = cell_start; i < cell_stop; ++i)
+            {
+                // get polar components of the velocity vector
+                auto u_mag = std::sqrt(
```

Comment:
This is ok and doesn't need to the changed, but elsewhere in WarpX, `auto` is typically only used when the type is explicitly specified in the RHS (or would be a long type definition).
Also, this and other variables here could be declared `const`.